docs/functions.adoc

```diff
@@ -0,0 +1,2750 @@
+  RETURN CAST_CHAR ( VALUE ) ;
+
+
+-- [cast_string] ---------
+
+CREATE FUNCTION "CAST_STRING" (
+  VALUE BOOL )
+  RETURNS STRING
+  SPECIFIC CAST_STRING__BOOL__STRING
+  RETURN CAST_STRING ( VALUE ) ;
+
+CREATE FUNCTION "CAST_STRING" (
+  VALUE CHAR )
+  RETURNS STRING
+  SPECIFIC CAST_STRING__CHAR__STRING
+  RETURN CAST_STRING ( VALUE ) ;
+
+CREATE FUNCTION "CAST_STRING" (
+  VALUE STRING )
+  RETURNS STRING
+  SPECIFIC CAST_STRING__STRING__STRING
+  RETURN CAST_STRING ( VALUE ) ;
+
+CREATE FUNCTION "CAST_STRING" (
+  VALUE SYMBOL )
+  RETURNS STRING
+  SPECIFIC CAST_STRING__SYMBOL__STRING
+  RETURN CAST_STRING ( VALUE ) ;
+
+
+-- [cast_symbol] ---------
+
+CREATE FUNCTION "CAST_SYMBOL" (
+  VALUE BOOL )
+  RETURNS SYMBOL
+  SPECIFIC CAST_SYMBOL__BOOL__SYMBOL
+  RETURN CAST_SYMBOL ( VALUE ) ;
+
+CREATE FUNCTION "CAST_SYMBOL" (
+  VALUE CHAR )
+  RETURNS SYMBOL
+  SPECIFIC CAST_SYMBOL__CHAR__SYMBOL
+  RETURN CAST_SYMBOL ( VALUE ) ;
+
+CREATE FUNCTION "CAST_SYMBOL" (
+  VALUE STRING )
+  RETURNS SYMBOL
+  SPECIFIC CAST_SYMBOL__STRING__SYMBOL
+  RETURN CAST_SYMBOL ( VALUE ) ;
+
+CREATE FUNCTION "CAST_SYMBOL" (
+  VALUE SYMBOL )
+  RETURNS SYMBOL
+  SPECIFIC CAST_SYMBOL__SYMBOL__SYMBOL
+  RETURN CAST_SYMBOL ( VALUE ) ;
+
+
+-- [cast_clob] ---------
+
+CREATE FUNCTION "CAST_CLOB" (
+  VALUE STRING )
+  RETURNS CLOB
+  SPECIFIC CAST_CLOB__STRING__CLOB
+  RETURN CAST_CLOB ( VALUE ) ;
+
+CREATE FUNCTION "CAST_CLOB" (
+  VALUE CLOB )
+  RETURNS CLOB
+  SPECIFIC CAST_CLOB__CLOB__CLOB
+  RETURN CAST_CLOB ( VALUE ) ;
+
+CREATE FUNCTION "CAST_CLOB" (
+  VALUE SYMBOL )
+  RETURNS CLOB
+  SPECIFIC CAST_CLOB__SYMBOL__CLOB
+  RETURN CAST_CLOB ( VALUE ) ;
+
+
+-- [cast_bag] ---------
+
+CREATE FUNCTION "CAST_BAG" (
+  VALUE BAG )
+  RETURNS BAG
+  SPECIFIC CAST_BAG__BAG__BAG
+  RETURN CAST_BAG ( VALUE ) ;
+
+CREATE FUNCTION "CAST_BAG" (
+  VALUE LIST )
+  RETURNS BAG
+  SPECIFIC CAST_BAG__LIST__BAG
+  RETURN CAST_BAG ( VALUE ) ;
+
+CREATE FUNCTION "CAST_BAG" (
+  VALUE SEXP )
+  RETURNS BAG
+  SPECIFIC CAST_BAG__SEXP__BAG
+  RETURN CAST_BAG ( VALUE ) ;
+
+
+-- [cast_list] ---------
+
+CREATE FUNCTION "CAST_LIST" (
+  VALUE LIST )
+  RETURNS LIST
+  SPECIFIC CAST_LIST__LIST__LIST
+  RETURN CAST_LIST ( VALUE ) ;
+
+CREATE FUNCTION "CAST_LIST" (
+  VALUE SEXP )
+  RETURNS LIST
+  SPECIFIC CAST_LIST__SEXP__LIST
+  RETURN CAST_LIST ( VALUE ) ;
+
+
+-- [cast_sexp] ---------
+
+CREATE FUNCTION "CAST_SEXP" (
+  VALUE LIST )
+  RETURNS SEXP
+  SPECIFIC CAST_SEXP__LIST__SEXP
+  RETURN CAST_SEXP ( VALUE ) ;
+
+CREATE FUNCTION "CAST_SEXP" (
+  VALUE SEXP )
+  RETURNS SEXP
+  SPECIFIC CAST_SEXP__SEXP__SEXP
+  RETURN CAST_SEXP ( VALUE ) ;
+
+
+-- [cast_struct] ---------
+
+CREATE FUNCTION "CAST_STRUCT" (
+  VALUE STRUCT )
+  RETURNS STRUCT
+  SPECIFIC CAST_STRUCT__STRUCT__STRUCT
+  RETURN CAST_STRUCT ( VALUE ) ;
+
+
+-- [not] ---------
+
+CREATE FUNCTION "NOT" (
+  VALUE BOOL )
+  RETURNS BOOL
+  SPECIFIC NOT__BOOL__BOOL
+  RETURN NOT ( VALUE ) ;
```

Comment:
I believe for all of these functions (logical, arithmetic, etc), we need to also add the routine characteristics of:
- `RETURNS NULL ON NULL INPUT`
- `DETERMINISTIC`

Because, for both of these, the opposite is implicit.